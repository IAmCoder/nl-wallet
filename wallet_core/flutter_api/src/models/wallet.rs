use serde::{Deserialize, Serialize};

use wallet::wallet::WalletUnlockError;

impl From<Result<(), WalletUnlockError>> for WalletUnlockResult {
    fn from(source: Result<(), WalletUnlockError>) -> Self {
        match source {
            Ok(()) => WalletUnlockResult::Ok,
            Err(err) => match err {
                WalletUnlockError::IncorrectPin {
                    leftover_attempts,
                    is_final_attempt,
                } => WalletUnlockResult::IncorrectPin {
                    leftover_attempts,
                    is_final_attempt,
                },
                WalletUnlockError::ServerError => WalletUnlockResult::ServerError,
                WalletUnlockError::Timeout { timeout_millis } => WalletUnlockResult::Timeout { timeout_millis },
                WalletUnlockError::Blocked => WalletUnlockResult::Blocked,
            },
        }
    }
}

#[derive(Debug, Deserialize, Serialize)]
pub enum WalletUnlockResult {
    Ok,
    IncorrectPin {
        leftover_attempts: u8,
        is_final_attempt: bool,
    },
    Timeout {
        timeout_millis: u32,
    },
    Blocked,
    ServerError,
}
