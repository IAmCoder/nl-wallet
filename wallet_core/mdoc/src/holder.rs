use coset::{iana, CoseMac0Builder, CoseSign1Builder, HeaderBuilder};
use ecdsa::SigningKey;
use ecdsa::{elliptic_curve::rand_core::OsRng, signature::Signer};
use indexmap::IndexMap;
use p256::NistP256;
use serde_bytes::ByteBuf;

use x509_parser::nom::AsBytes;
use x509_parser::prelude::X509Certificate;

use crate::serialization::{cbor_deserialize, CborError};
use crate::Error;
use crate::{
    basic_sa_ext::{
        DataToIssueMessage, Entry, KeyGenerationResponseMessage, RequestKeyGenerationMessage, SparseIssuerSigned,
        UnsignedMdoc,
    },
    cose::ClonePayload,
    crypto::{dh_hmac_key, sha256},
    iso::*,
    serialization::{cbor_serialize, TaggedBytes},
    verifier::X509Subject,
    Result,
};

#[derive(thiserror::Error, Debug)]
pub enum HolderError {
    #[error("unsatisfiable request: DocType {0} not in wallet")]
    UnsatisfiableRequest(DocType),
    #[error("readerAuth not present for all documents")]
    ReaderAuthMissing,
    #[error("document requests were signed by different readers")]
    ReaderAuthsInconsistent,
    #[error(transparent)]
    Cbor(#[from] CborError),
}

/// Mdoc credentials of the holder. This data structure supports storing:
/// - mdocs with different doctypes, through the map over `DocType`,
/// - multiple mdocs having the same doctype but distinct attributes, through the map over `Vec<u8>` which is computed
///   with [`Credential::hash()`] (see its rustdoc for details),
/// - multiple mdocs having the same doctype and the same attributes, through the `CredentialCopies` data structure.
#[derive(Debug, Clone)]
pub struct Credentials(pub IndexMap<DocType, IndexMap<Vec<u8>, CredentialCopies>>);

impl Default for Credentials {
    fn default() -> Self {
        Self::new()
    }
}
impl<const N: usize> TryFrom<[Credential; N]> for Credentials {
    type Error = Error;

    fn try_from(value: [Credential; N]) -> Result<Self> {
        let mut creds = Credentials::new();
        creds.add(value.into_iter())?;
        Ok(creds)
    }
}

#[derive(Debug)]
pub struct IssuanceState<'a> {
    pub request: &'a RequestKeyGenerationMessage,
    pub response: KeyGenerationResponseMessage,

    /// Private keys grouped by distinct credentials, and then per copies of each distinct credential.
    pub private_keys: Vec<Vec<SigningKey<NistP256>>>,
}

impl Credentials {
    pub fn new() -> Credentials {
        Credentials(IndexMap::new())
    }

    pub fn add(&mut self, creds: impl Iterator<Item = Credential>) -> Result<()> {
        for cred in creds.into_iter() {
            self.0
                .entry(cred.doc_type.clone())
                .or_insert(IndexMap::new())
                .entry(cred.hash()?)
                .or_insert(CredentialCopies::new())
                .creds
                .push(cred);
        }

        Ok(())
    }

    fn generate_keys(count: u64) -> Vec<SigningKey<p256::NistP256>> {
        (0..count)
            .map(|_| SigningKey::<p256::NistP256>::random(OsRng))
            .collect()
    }

    pub fn issuance_start(request: &RequestKeyGenerationMessage) -> Result<IssuanceState> {
        let private_keys = request
            .unsigned_mdocs
            .iter()
            .map(|unsigned| Credentials::generate_keys(unsigned.count))
            .collect::<Vec<_>>();
        let response = KeyGenerationResponseMessage::new(request, &private_keys)?;

        Ok(IssuanceState {
            request,
            private_keys,
            response,
        })
    }

    pub fn issuance_finish(
        state: IssuanceState,
        issuer_response: DataToIssueMessage,
        issuer_cert: &X509Certificate,
    ) -> Result<Vec<CredentialCopies>> {
        issuer_response
            .mobile_id_documents
            .iter()
            .zip(&state.request.unsigned_mdocs)
            .zip(&state.private_keys)
            .map(|((doc, unsigned), keys)| {
                Ok(doc
                    .sparse_issuer_signed
                    .iter()
                    .zip(keys)
                    .map(|(iss_signature, key)| iss_signature.to_credential(key.clone(), unsigned, issuer_cert))
                    .collect::<Result<Vec<_>>>()?
                    .into())
            })
            .collect()
    }

    pub fn disclose(&self, device_request: &DeviceRequest, challenge: &[u8]) -> Result<DeviceResponse> {
        let mut docs: Vec<Document> = Vec::new();

        for doc_request in &device_request.doc_requests {
            let items_request = &doc_request.items_request.0;
            if !self.0.contains_key(&items_request.doc_type) {
                return Err(HolderError::UnsatisfiableRequest(items_request.doc_type.clone()).into());
            }

            // This takes any mdoc of the specified doctype. TODO: allow user choice.
            let cred = &self.0[&items_request.doc_type].values().next().unwrap().creds[0];
            docs.push(cred.disclose_document(items_request, challenge)?);
        }

        Ok(DeviceResponse {
            version: "1.0".to_string(),
            documents: Some(docs),
            document_errors: None,
            status: 0,
        })
    }
}

/// Stores multiple copies of mdocs that have identical attributes.
///
/// TODO: support marking an mdoc has having been used, so that it can be avoided in future disclosures,
/// for unlinkability.
#[derive(Debug, Clone)]
pub struct CredentialCopies {
    pub creds: Vec<Credential>,
}

impl IntoIterator for CredentialCopies {
    type Item = Credential;
    type IntoIter = std::vec::IntoIter<Credential>;
    fn into_iter(self) -> Self::IntoIter {
        self.creds.into_iter()
    }
}
impl From<Vec<Credential>> for CredentialCopies {
    fn from(creds: Vec<Credential>) -> Self {
        Self { creds }
    }
}
impl CredentialCopies {
    fn new() -> Self {
        CredentialCopies { creds: vec![] }
    }
}

/// A full mdoc credential: everything needed to disclose attributes from the mdoc.
#[derive(Debug, Clone)]
pub struct Credential {
    private_key: ecdsa::SigningKey<p256::NistP256>,
    issuer_signed: IssuerSigned,
    doc_type: String,
}

impl Credential {
    pub fn new(
        private_key: ecdsa::SigningKey<p256::NistP256>,
        issuer_signed: IssuerSigned,
        ca_cert: &X509Certificate,
    ) -> Result<Credential> {
        let (_, mso) = issuer_signed.verify(ca_cert)?;
        Ok(Credential {
            private_key,
            issuer_signed,
            doc_type: mso.doc_type,
        })
    }

    /// Hash of the credential, acting as an identifier for the credential that takes into account its doctype
    /// and all of its attributes. Computed schematically as `SHA256(CBOR(doctype, attributes))`.
    fn hash(&self) -> Result<Vec<u8>> {
        Ok(sha256(
            &cbor_serialize(&(
                &self.doc_type,
                &self
                    .issuer_signed
                    .name_spaces
                    .as_ref()
                    .unwrap()
                    .iter()
                    .map(|(namespace, attrs)| (namespace.clone(), Vec::<Entry>::from(attrs)))
                    .collect::<IndexMap<_, _>>(),
            ))
            .map_err(HolderError::Cbor)?,
        ))
    }

    pub fn disclose_document(&self, items_request: &ItemsRequest, challenge: &[u8]) -> Result<Document> {
        let disclosed_namespaces: IssuerNameSpaces = self
            .issuer_signed
            .name_spaces
            .as_ref()
            .unwrap()
            .iter()
            .filter(|&(namespace, _)| items_request.name_spaces.contains_key(namespace))
            .map(|(namespace, attributes)| {
                (
                    namespace.clone(),
                    attributes.filter(items_request.name_spaces.get(namespace).unwrap()),
                )
            })
            .collect();

        Ok(Document {
            doc_type: items_request.doc_type.clone(),
            issuer_signed: IssuerSigned {
                name_spaces: Some(disclosed_namespaces),
                issuer_auth: self.issuer_signed.issuer_auth.clone(),
            },
            device_signed: DeviceSigned::new_signature(&self.private_key, challenge),
            errors: None,
        })
    }
}

impl SparseIssuerSigned {
    fn to_credential(
        &self,
        private_key: SigningKey<p256::NistP256>,
        unsigned: &UnsignedMdoc,
        iss_cert: &X509Certificate,
    ) -> Result<Credential> {
        let name_spaces: IssuerNameSpaces = unsigned
            .attributes
            .iter()
            .map(|(namespace, attrs)| {
                (
                    namespace.clone(),
                    attrs
                        .iter()
                        .enumerate()
                        .map(|(index, attr)| attr.to_issuer_signed_item(index, self.randoms[namespace][index].to_vec()))
                        .collect::<Vec<_>>()
                        .into(),
                )
            })
            .collect();

        let mso = MobileSecurityObject {
            version: self.sparse_issuer_auth.version.clone(),
            digest_algorithm: self.sparse_issuer_auth.digest_algorithm.clone(),
            value_digests: (&name_spaces).try_into()?,
            device_key_info: private_key.verifying_key().try_into()?,
            doc_type: unsigned.doc_type.clone(),
            validity_info: self.sparse_issuer_auth.validity_info.clone(),
        };
        let issuer_auth = self
            .sparse_issuer_auth
            .issuer_auth
            .clone_with_payload(cbor_serialize(&TaggedBytes::from(mso)).map_err(HolderError::Cbor)?);

        let issuer_signed = IssuerSigned {
            name_spaces: Some(name_spaces),
            issuer_auth,
        };
        issuer_signed.verify(iss_cert)?;

        Ok(Credential {
            private_key,
            issuer_signed,
            doc_type: unsigned.doc_type.clone(),
        })
    }
}

impl Entry {
    fn to_issuer_signed_item(&self, index: usize, random: Vec<u8>) -> IssuerSignedItemBytes {
        IssuerSignedItem {
            digest_id: index as u64,
            random: ByteBuf::from(random),
            element_identifier: self.name.clone(),
            element_value: self.value.clone(),
        }
        .into()
    }
}

impl Attributes {
    /// Return a copy that contains only the items requested in `items_request`.
    fn filter(&self, requested: &DataElements) -> Attributes {
        self.0
            .clone()
            .into_iter()
            .filter(|attr| requested.contains_key(&attr.0.element_identifier))
            .collect::<Vec<_>>()
            .into()
    }
}

impl DeviceSigned {
    pub fn new_signature(private_key: &ecdsa::SigningKey<p256::NistP256>, challenge: &[u8]) -> DeviceSigned {
        let cose = CoseSign1Builder::new()
            .payload(Vec::from(challenge))
            .protected(HeaderBuilder::new().algorithm(iana::Algorithm::ES256).build())
            .create_signature(&[], |data| private_key.sign(data).to_vec())
            .build()
            .clone_without_payload();

        DeviceSigned {
            name_spaces: IndexMap::new().into(),
            device_auth: DeviceAuth::DeviceSignature(cose.into()),
        }
    }

    #[allow(dead_code)] // TODO test this
    pub fn new_mac(
        private_key: &ecdsa::SigningKey<p256::NistP256>,
        reader_pub_key: &ecdsa::VerifyingKey<p256::NistP256>,
        challenge: &[u8],
    ) -> Result<DeviceSigned> {
        let device_auth: DeviceAuthenticationBytes = cbor_deserialize(challenge).map_err(HolderError::Cbor)?;
        let key = dh_hmac_key(
            private_key,
            reader_pub_key,
            device_auth.0.session_transcript_bts()?.as_bytes(),
            "EMacKey",
            32,
        )?;

        let cose = CoseMac0Builder::new()
            .payload(Vec::from(challenge))
            .protected(HeaderBuilder::new().algorithm(iana::Algorithm::ES256).build())
            .create_tag(&[], |data| ring::hmac::sign(&key, data).as_ref().into())
            .build()
            .clone_without_payload();

        Ok(DeviceSigned {
            name_spaces: IndexMap::new().into(),
            device_auth: DeviceAuth::DeviceMac(cose.into()),
        })
    }
}

impl DeviceRequest {
    /// Verify reader authentication, if present.
    /// Note that since each DocRequest carries its own reader authentication, the spec allows the
    /// the DocRequests to be signed by distinct readers. TODO maybe support this.
    /// For now, this function requires either none of the DocRequests to be signed, or all of them
    /// by the same reader.
    #[allow(dead_code)] // TODO use in client
    pub fn verify(&self, ca_cert: &X509Certificate, reader_authentication_bts: &[u8]) -> Result<Option<X509Subject>> {
        if self.doc_requests.iter().all(|d| d.reader_auth.is_none()) {
            return Ok(None);
        }
        if self.doc_requests.iter().any(|d| d.reader_auth.is_none()) {
            return Err(HolderError::ReaderAuthMissing.into());
        }

        let mut reader: Option<X509Subject> = None;
        for doc_request in &self.doc_requests {
            let (_, found) = doc_request
                .reader_auth
                .as_ref()
                .unwrap()
                .clone_with_payload(reader_authentication_bts.to_vec())
                .verify_against_cert(ca_cert)?;
            if reader.is_none() {
                reader.replace(found);
            } else if *reader.as_ref().unwrap() != found {
                return Err(HolderError::ReaderAuthsInconsistent.into());
            }
        }

        Ok(reader)
    }
}
