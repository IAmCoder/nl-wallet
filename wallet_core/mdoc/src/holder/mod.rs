//! Holder software, containing a [`Wallet`] that can store, receive, and disclose mdocs.
//! See [`MdocRetriever`], [`Wallet::start_issuance()`], and [`Wallet::disclose()`] respectively.

use std::error::Error;

use crate::{
    identifiers::AttributeIdentifier,
    iso::*,
    utils::{
        reader_auth::{self, ReaderRegistration},
        x509::{Certificate, CertificateError},
    },
};

pub mod disclosure;
pub use disclosure::*;

pub mod issuance;
pub use issuance::*;

pub mod mdocs;
pub use mdocs::*;

#[derive(thiserror::Error, Debug)]
pub enum HolderError {
    #[error("unsatisfiable request: DocType {0} not in wallet")]
    UnsatisfiableRequest(DocType),
    #[error("readerAuth not present for all documents")]
    ReaderAuthMissing,
    #[error("document requests were signed by different readers")]
    ReaderAuthsInconsistent,
    #[error("issuer not trusted for doctype {0}")]
    UntrustedIssuer(DocType),
    #[error("certificate error: {0}")]
    CertificateError(#[from] CertificateError),
    #[error("wrong private key type")]
    PrivateKeyTypeMismatch { expected: String, have: String },
    #[error("request error: {0}")]
    RequestError(#[from] reqwest::Error),
    #[error("malformed Service Engagement: url missing")]
    MalformedServiceEngagement,
    #[error("malformed attribute: random too short (was {0}; minimum {1}")]
    AttributeRandomLength(usize, usize),
    #[error("missing issuance session state")]
    MissingIssuanceSessionState,
    #[error("verifier URL not present in reader engagement")]
    VerifierUrlMissing,
    #[error("verifier ephemeral key not present in reader engagement")]
    VerifierEphemeralKeyMissing,
    #[error("no document requests are present in device request")]
    NoAttributesRequested,
    #[error("no reader registration present in certificate")]
    NoReaderRegistration(Certificate),
    #[error("reader registration attribute validation failed: {0}")]
    ReaderRegistrationValidation(#[from] reader_auth::ValidationError),
    #[error("could not retrieve docs from source: {0}")]
    MdocDataSource(#[source] Box<dyn Error + Send + Sync>),
    #[error("multiple candidates for disclosure is unsupported, found for doc types: {}", .0.join(", "))]
    MultipleCandidates(Vec<DocType>),
    #[error("not all requested attributes are available, missing: {missing_attributes:?}")]
    AttributesNotAvailable {
        reader_registration: Box<ReaderRegistration>,
        missing_attributes: Vec<AttributeIdentifier>,
    },
}
