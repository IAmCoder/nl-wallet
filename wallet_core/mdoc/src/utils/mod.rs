pub mod cose;
pub mod keys;
pub mod reader_auth;
pub mod serialization;
pub mod x509;

pub(crate) mod crypto;

#[cfg(feature = "mock")]
pub mod mdocs_map;
