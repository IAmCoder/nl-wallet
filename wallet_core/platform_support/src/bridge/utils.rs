use std::fmt::Debug;

use super::get_platform_support;

pub use crate::utils::UtilitiesError;

// this is required to catch UnexpectedUniFFICallbackError
impl From<uniffi::UnexpectedUniFFICallbackError> for UtilitiesError {
    fn from(value: uniffi::UnexpectedUniFFICallbackError) -> Self {
        Self::BridgingError { reason: value.reason }
    }
}

// the callback traits defined in the UDL, which we have write out here ourselves
pub trait UtilitiesBridge: Send + Sync + Debug {
    fn get_storage_path(&self) -> Result<String, UtilitiesError>;
}

pub fn get_utils_bridge() -> &'static dyn UtilitiesBridge {
    get_platform_support().utils.as_ref()
}
