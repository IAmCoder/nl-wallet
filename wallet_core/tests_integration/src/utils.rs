use std::env;
use std::path::PathBuf;

use wallet_common::urls::BaseUrl;

pub(crate) fn path_from_root(file_name: &str) -> PathBuf {
    env::var("CARGO_MANIFEST_DIR")
        .map(PathBuf::from)
        .unwrap_or_default()
        .join(file_name)
}

pub(crate) fn read_file(file_name: &str) -> Vec<u8> {
    std::fs::read(path_from_root(file_name)).unwrap()
}

pub(crate) fn remove_path(base_url: &BaseUrl) -> BaseUrl {
    let mut base_url = base_url.as_ref().clone();
    base_url.set_path("");
    base_url.try_into().unwrap()
}
