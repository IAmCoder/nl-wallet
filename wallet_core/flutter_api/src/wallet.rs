use once_cell::sync::Lazy;
use tokio::sync::Mutex;
use wallet::{init_wallet, Wallet};

pub static WALLET: Lazy<Mutex<Wallet>> = Lazy::new(|| Mutex::new(init_wallet()));
