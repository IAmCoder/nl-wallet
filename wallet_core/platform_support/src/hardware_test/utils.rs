use crate::utils::{hardware::HardwareUtilities, test};

// this is the starting point for the integration test performed from Android / iOS.
#[no_mangle]
fn utils_test_get_storage_path() -> bool {
    let rt = tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap();

    rt.block_on(test::get_and_verify_storage_path::<HardwareUtilities>())
}

#[cfg(target_os = "android")]
mod android {
    use jni::{objects::JClass, JNIEnv};

    #[rustfmt::skip]
    #[no_mangle]
    extern "C" fn Java_nl_rijksoverheid_edi_wallet_platform_1support_utilities_UtilitiesBridgeInstrumentedTest_utilities_1test_1storage_1path(
        _env: JNIEnv,
        _: JClass,
    ) -> bool {
        super::utils_test_get_storage_path()
    }
}
