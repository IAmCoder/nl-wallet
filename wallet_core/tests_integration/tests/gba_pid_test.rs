use indexmap::IndexMap;
use rstest::rstest;
use tokio::fs;
use uuid::Uuid;

use nl_wallet_mdoc::holder::{CborHttpClient, DisclosureSession};
use openid4vc::{
    issuance_session::{HttpIssuanceSession, IssuanceSessionError},
    ErrorResponse,
};
use platform_support::utils::{software::SoftwareUtilities, PlatformUtilities};
use tests_integration::fake_digid::fake_digid_auth;
use wallet::{
    errors::PidIssuanceError,
    mock::{default_configuration, MockStorage},
    wallet_deps::{
        ConfigServerConfiguration, ConfigurationRepository, HttpAccountProviderClient, HttpConfigurationRepository,
        HttpDigidSession, UpdateableConfigurationRepository,
    },
    Wallet,
};
use wallet_common::keys::software::SoftwareEcdsaKey;

#[tokio::test]
#[rstest]
#[should_panic(expected = "conversion error")]
async fn test_gba_pid_conversion_error(
    #[values("999993318", "999993896", "999990585", "999991127", "999992326", "999997658")] bsn: &str,
) {
    gba_pid(bsn).await
}

#[tokio::test]
#[rstest]
#[should_panic(expected = "unknown bsn")]
async fn test_gba_pid_unknown_bsn(#[values("999992306", "999995657")] bsn: &str) {
    gba_pid(bsn).await
}

#[tokio::test]
#[rstest]
async fn test_gba_pid_success(
    #[values(
        "999991772",
        "999991000",
        "999992958",
        "999991838",
        "999991802",
        "999991644",
        "999994761",
        "999990159",
        "999993598",
        "000009842",
        "999991747",
        "999994785",
        "999992636",
        "999993811",
        "999990640",
        "999992107",
        "999992120",
        "999991577",
        "999992533",
        "999994931",
        "999993215",
        "999992983",
        "999990044",
        "999990196",
        "999993446",
        "999992880",
        "000009878",
        "999991516",
        "999991292",
        "999991401",
        "999992569",
        "999991814",
        "999994359",
        "999994542",
        "999990160",
        "999992065",
        "999991565",
        "999991243",
        "999990871",
        "999990500",
        "999993665",
        "999990627",
        "999993409",
        "999997634",
        "999997646",
        "999997671",
        "999997683",
        "999997695",
        "999997701",
        "999997713",
        "999997725",
        "999997737",
        "999997749",
        "999997750",
        "010245741",
        "010755561",
        "999998341",
        "999998353"
    )]
    bsn: &str,
) {
    gba_pid(bsn).await
}

async fn gba_pid(bsn: &str) {
    let storage_path = SoftwareUtilities::storage_path().await.unwrap();
    let etag_file = storage_path.join("latest-configuration-etag.txt");
    // make sure there are no storage files from previous test runs
    let _ = fs::remove_file(etag_file.as_path()).await;

    let config_server_config = ConfigServerConfiguration::default();
    let wallet_config = default_configuration();

    let config_repository = HttpConfigurationRepository::new(
        config_server_config.base_url,
        config_server_config.trust_anchors,
        (&config_server_config.signing_public_key).into(),
        storage_path,
        wallet_config,
    )
    .await
    .unwrap();
    config_repository.fetch().await.unwrap();
    let pid_issuance_config = &config_repository.config().pid_issuance;

    let mut wallet: Wallet<
        HttpConfigurationRepository,
        MockStorage,
        SoftwareEcdsaKey,
        HttpAccountProviderClient,
        HttpDigidSession,
        HttpIssuanceSession,
        DisclosureSession<CborHttpClient, Uuid>,
    > = Wallet::init_registration(
        config_repository,
        MockStorage::default(),
        HttpAccountProviderClient::default(),
    )
    .await
    .expect("Could not create test wallet");

    let pin = String::from("123344");

    wallet.register(pin.clone()).await.expect("Could not register wallet");

    let authorization_url = wallet
        .create_pid_issuance_auth_url()
        .await
        .expect("Could not create pid issuance auth url");

    let redirect_url = fake_digid_auth(
        &authorization_url,
        &pid_issuance_config.digid_url,
        pid_issuance_config.digid_trust_anchors(),
        bsn,
    )
    .await;

    let unsigned_mdocs_result = wallet.continue_pid_issuance(redirect_url).await;
    let unsigned_mdocs = match unsigned_mdocs_result {
        Ok(mdocs) => mdocs,
        Err(PidIssuanceError::PidIssuer(IssuanceSessionError::TokenRequest(ErrorResponse {
            error_description: Some(description),
            ..
        }))) if description.contains("Error converting GBA-V XML to Haal-Centraal JSON: GBA-V error") => {
            panic!("conversion error")
        }
        Err(PidIssuanceError::PidIssuer(IssuanceSessionError::TokenRequest(ErrorResponse {
            error_description: Some(description),
            ..
        }))) if description.contains("could not find attributes for BSN") => panic!("unknown bsn"),
        Err(e) => {
            dbg!("{:?}", e);
            panic!("could not continue pid issuance")
        }
    };

    let attributes = unsigned_mdocs.into_iter().fold(IndexMap::new(), |mut attrs, mdoc| {
        mdoc.attributes.into_iter().for_each(|(key, attr)| {
            attrs.insert(key, attr.value);
        });
        attrs
    });

    insta::with_settings!({
        description => format!("BSN: {}", bsn),
        snapshot_suffix => bsn,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_yaml_snapshot!(attributes);
    });
}
