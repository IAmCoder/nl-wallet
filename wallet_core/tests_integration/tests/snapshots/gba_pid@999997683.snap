---
source: tests_integration/tests/gba_pid_test.rs
assertion_line: 174
description: "BSN: 999997683"
expression: attributes
---
com.example.pid__given_name:
  String: MariÃ«lla
com.example.pid__family_name:
  String: Janssen
com.example.pid__own_family_name:
  String: Janssen
com.example.pid__gender:
  Gender: Female
com.example.pid__birth_date:
  Date: 1972-10-24
com.example.pid__age_over_18:
  Boolean: true
com.example.pid__has_spouse_or_partner:
  Boolean: false
com.example.pid__bsn:
  String: "999997683"
com.example.address__resident_street:
  String: Thorbeckelaan
com.example.address__resident_house_number:
  String: "16"
com.example.address__resident_postal_code:
  String: 3552CS
com.example.address__resident_city:
  String: Utrecht
com.example.address__resident_country:
  String: Nederland
